File: apps/mobile_android/lib/goals.dart
import 'package:core_models/core_models.dart';

/// How wide a net the goal covers.
enum GoalPeriod { week, month }

/// A metric a [RunGoal] can track. A single goal can have any subset of
/// these active — the dashboard card shows one progress row per active
/// target, so "20 km, 5 runs, 5:00/km average" is a single goal with three
/// targets, not three separate goals.
enum GoalTargetKind { distance, time, avgPace, runCount }

/// Human-readable label used in the editor and dashboard card.
String goalKindLabel(GoalTargetKind kind) {
  return switch (kind) {
    GoalTargetKind.distance => 'Distance',
    GoalTargetKind.time => 'Time',
    GoalTargetKind.avgPace => 'Avg pace',
    GoalTargetKind.runCount => 'Runs',
  };
}

/// A user-defined training goal. One goal holds the period plus zero or
/// more concrete targets (distance / time / avg pace / run count). Stored
/// locally in [Preferences]; never round-tripped to Supabase.
class RunGoal {
  final String id;
  final GoalPeriod period;

  /// Optional display name. Null means "auto-label from targets" — see
  /// [displayTitle]. Exists so a user with two goals in the same period
  /// can tell them apart ("Base miles" vs "Speed work").
  final String? title;

  final double? distanceMetres;
  final double? timeSeconds;
  final double? avgPaceSecPerKm;
  final double? runCount;

  const RunGoal({
    required this.id,
    required this.period,
    this.title,
    this.distanceMetres,
    this.timeSeconds,
    this.avgPaceSecPerKm,
    this.runCount,
  });

  /// The target kinds on this goal, in display order.
  List<GoalTargetKind> get activeKinds => [
        if (distanceMetres != null) GoalTargetKind.distance,
        if (timeSeconds != null) GoalTargetKind.time,
        if (avgPaceSecPerKm != null) GoalTargetKind.avgPace,
        if (runCount != null) GoalTargetKind.runCount,
      ];

  bool get isEmpty => activeKinds.isEmpty;

  Map<String, dynamic> toJson() => {
        'id': id,
        'period': period.name,
        if (title != null && title!.isNotEmpty) 'title': title,
        if (distanceMetres != null) 'distance_m': distanceMetres,
        if (timeSeconds != null) 'time_s': timeSeconds,
        if (avgPaceSecPerKm != null) 'pace_s_per_km': avgPaceSecPerKm,
        if (runCount != null) 'run_count': runCount,
      };

  factory RunGoal.fromJson(Map<String, dynamic> json) {
    final period = GoalPeriod.values.firstWhere(
      (p) => p.name == json['period'],
      orElse: () => GoalPeriod.week,
    );

    // Data migration: the first goals build used a single-target shape
    // `{type, period, target}`. Detect it by the presence of 'type' and
    // splat the value into the matching optional field. The next
    // [_persistGoals] writes the new shape back out, so this branch only
    // fires on the upgrade boot.
    if (json.containsKey('type')) {
      final type = json['type'] as String;
      final target = (json['target'] as num).toDouble();
      return RunGoal(
        id: json['id'] as String,
        period: period,
        distanceMetres: type == 'distance' ? target : null,
        timeSeconds: type == 'time' ? target : null,
        avgPaceSecPerKm: type == 'avgPace' ? target : null,
        runCount: type == 'runCount' ? target : null,
      );
    }

    final rawTitle = json['title'] as String?;
    return RunGoal(
      id: json['id'] as String,
      period: period,
      title: (rawTitle != null && rawTitle.isNotEmpty) ? rawTitle : null,
      distanceMetres: (json['distance_m'] as num?)?.toDouble(),
      timeSeconds: (json['time_s'] as num?)?.toDouble(),
      avgPaceSecPerKm: (json['pace_s_per_km'] as num?)?.toDouble(),
      runCount: (json['run_count'] as num?)?.toDouble(),
    );
  }
}

/// Generate an opaque id for a new goal. Not globally unique — fine for a
/// local-only list where collisions at microsecond resolution are impossible
/// in practice.
String newGoalId() => DateTime.now().microsecondsSinceEpoch.toRadixString(36);

/// Progress on a single target within a goal.
class TargetProgress {
  final GoalTargetKind kind;

  /// Current value in canonical units (metres / seconds / sec-per-km / count).
  final double current;

  /// Copy of the target in the same canonical units.
  final double target;

  /// Progress fraction in `[0, 1]`. For lower-is-better targets (avg pace)
  /// this is `target / current` clamped.
  final double percent;
  final bool complete;
  final String feedback;

  const TargetProgress({
    required this.kind,
    required this.current,
    required this.target,
    required this.percent,
    required this.complete,
    required this.feedback,
  });
}

/// Snapshot of how a goal is tracking. One [TargetProgress] per active
/// target on the [RunGoal], plus an aggregate [overallPercent] and a
/// single [complete] flag that's true only when every target is met.
class GoalProgress {
  final List<TargetProgress> targets;
  final double overallPercent;
  final bool complete;
  final int runCount;

  const GoalProgress({
    required this.targets,
    required this.overallPercent,
    required this.complete,
    required this.runCount,
  });
}

/// Pure evaluator: given a goal and the full run list, compute progress
/// for every active target.
GoalProgress evaluateGoal(RunGoal goal, List<Run> runs, DateTime now) {
  final periodStart = goalPeriodStart(goal.period, now);
  final periodEnd = goalPeriodEnd(goal.period, now);

  final inPeriod = runs
      .where((r) =>
          !r.startedAt.isBefore(periodStart) &&
          r.startedAt.isBefore(periodEnd))
      .toList();

  // Pace calculations exclude cycling — a distance-weighted average would
  // otherwise be dominated by a single long bike ride.
  final paceEligible = inPeriod
      .where((r) => r.metadata?['activity_type'] != 'cycle')
      .toList();

  final totalMetres = inPeriod.fold<double>(0, (s, r) => s + r.distanceMetres);
  final totalSeconds =
      inPeriod.fold<int>(0, (s, r) => s + r.duration.inSeconds);

  final targets = <TargetProgress>[];

  if (goal.distanceMetres != null) {
    targets.add(_evalCumulative(
      kind: GoalTargetKind.distance,
      target: goal.distanceMetres!,
      current: totalMetres,
      runsInPeriod: inPeriod.length,
      now: now,
      periodStart: periodStart,
      periodEnd: periodEnd,
      format: (m) => '${(m / 1000).toStringAsFixed(1)} km',
    ));
  }

  if (goal.timeSeconds != null) {
    targets.add(_evalCumulative(
      kind: GoalTargetKind.time,
      target: goal.timeSeconds!,
      current: totalSeconds.toDouble(),
      runsInPeriod: inPeriod.length,
      now: now,
      periodStart: periodStart,
      periodEnd: periodEnd,
      format: _formatSecondsCoarse,
    ));
  }

  if (goal.avgPaceSecPerKm != null) {
    final paceMetres =
        paceEligible.fold<double>(0, (s, r) => s + r.distanceMetres);
    final paceSecondsSum =
        paceEligible.fold<int>(0, (s, r) => s + r.duration.inSeconds);
    final current =
        paceMetres > 10 ? paceSecondsSum / (paceMetres / 1000) : 0.0;
    targets.add(_evalPace(
      target: goal.avgPaceSecPerKm!,
      current: current,
      runningRuns: paceEligible.length,
    ));
  }

  if (goal.runCount != null) {
    targets.add(_evalRunCount(
      target: goal.runCount!,
      current: inPeriod.length.toDouble(),
    ));
  }

  final overall = targets.isEmpty
      ? 0.0
      : targets.map((t) => t.percent).reduce((a, b) => a + b) / targets.length;
  final complete =
      targets.isNotEmpty && targets.every((t) => t.complete);

  return GoalProgress(
    targets: targets,
    overallPercent: overall,
    complete: complete,
    runCount: inPeriod.length,
  );
}

/// Progress for a higher-is-better cumulative target (distance, time).
TargetProgress _evalCumulative({
  required GoalTargetKind kind,
  required double target,
  required double current,
  required int runsInPeriod,
  required DateTime now,
  required DateTime periodStart,
  required DateTime periodEnd,
  required String Function(double) format,
}) {
  final percent = target > 0 ? (current / target).clamp(0.0, 1.0) : 0.0;
  final complete = target > 0 && current >= target;

  String feedback;
  if (runsInPeriod == 0) {
    feedback = 'Log a run to start tracking';
  } else if (complete) {
    feedback = 'Goal reached';
  } else {
    final totalSec = periodEnd.difference(periodStart).inSeconds.toDouble();
    final elapsedSec =
        now.difference(periodStart).inSeconds.clamp(0, totalSec.toInt());
    final expected = totalSec > 0 ? target * (elapsedSec / totalSec) : 0.0;
    final delta = current - expected;
    feedback = delta > 0
        ? '${format(delta)} ahead of pace'
        : '${format(target - current)} to go';
  }

  return TargetProgress(
    kind: kind,
    current: current,
    target: target,
    percent: percent,
    complete: complete,
    feedback: feedback,
  );
}

TargetProgress _evalPace({
  required double target,
  required double current,
  required int runningRuns,
}) {
  double percent;
  bool complete;
  String feedback;

  if (current <= 0) {
    percent = 0;
    complete = false;
    feedback = runningRuns == 0
        ? 'Log a running activity to track pace'
        : 'Log a run to start tracking';
  } else if (current <= target) {
    percent = 1.0;
    complete = true;
    feedback = 'Goal reached';
  } else {
    percent = (target / current).clamp(0.0, 1.0);
    complete = false;
    final delta = current - target;
    if (delta.abs() < 1) {
      feedback = 'On target';
    } else {
      feedback = '${delta.abs().round()}s off target';
    }
  }

  return TargetProgress(
    kind: GoalTargetKind.avgPace,
    current: current,
    target: target,
    percent: percent,
    complete: complete,
    feedback: feedback,
  );
}

TargetProgress _evalRunCount({
  required double target,
  required double current,
}) {
  final percent = target > 0 ? (current / target).clamp(0.0, 1.0) : 0.0;
  final complete = target > 0 && current >= target;

  String feedback;
  if (complete) {
    feedback = 'Goal reached';
  } else if (current == 0) {
    feedback = 'Log a run to start tracking';
  } else {
    final remaining = (target - current).ceil();
    feedback = '$remaining to go';
  }

  return TargetProgress(
    kind: GoalTargetKind.runCount,
    current: current,
    target: target,
    percent: percent,
    complete: complete,
    feedback: feedback,
  );
}

String _formatSecondsCoarse(double seconds) {
  final totalMin = (seconds / 60).round();
  if (totalMin >= 60) {
    final h = totalMin ~/ 60;
    final m = totalMin % 60;
    return m > 0 ? '${h}h ${m}m' : '${h}h';
  }
  return '${totalMin}m';
}

/// Monday 00:00 local time of the ISO week containing [now]. The single
/// source of truth for "this week" across goals, the history filter, and
/// the dashboard summary cards — keeping them in lockstep so a future
/// change of convention (e.g. Sunday-start) is a one-file edit.
DateTime weekStartLocal(DateTime now) {
  final startOfToday = DateTime(now.year, now.month, now.day);
  final daysFromMonday = (now.weekday - DateTime.monday) % 7;
  return startOfToday.subtract(Duration(days: daysFromMonday));
}

/// Start of the period containing [now], inclusive, in local time.
DateTime goalPeriodStart(GoalPeriod period, DateTime now) {
  switch (period) {
    case GoalPeriod.week:
      return weekStartLocal(now);
    case GoalPeriod.month:
      return DateTime(now.year, now.month, 1);
  }
}

/// Exclusive end of the period containing [now], in local time.
DateTime goalPeriodEnd(GoalPeriod period, DateTime now) {
  switch (period) {
    case GoalPeriod.week:
      return goalPeriodStart(period, now).add(const Duration(days: 7));
    case GoalPeriod.month:
      final nextMonth = now.month == 12 ? 1 : now.month + 1;
      final year = now.month == 12 ? now.year + 1 : now.year;
      return DateTime(year, nextMonth, 1);
  }
}

File: apps/web/src/lib/goals.ts
/// Multi-metric user goals, ported from `apps/mobile_android/lib/goals.dart`.
/// Four target kinds — distance, time, average pace, run count — over a
/// week or month period. Pace targets are distance-weighted and exclude
/// cycling (a single long bike ride would otherwise dominate the
/// average and make the metric meaningless for runners).
///
/// Stored in `localStorage` under a single JSON blob, mirroring the
/// local-first shape of the Android version. Not bag-synced today —
/// goals are highly personal and per-user, and the universal settings
/// bag is already carrying a scalar `weekly_mileage_goal_m` for the
/// simple case. When a user wants cross-device sync, promote this to
/// the bag as `run_goals` (array) with a new registered key.

import type { Run } from './types';

export type GoalPeriod = 'week' | 'month';

export interface RunGoal {
	id: string;
	period: GoalPeriod;
	title?: string;
	distanceMetres?: number;
	timeSeconds?: number;
	/// Lower-is-better target. Stored canonically as seconds per
	/// kilometre regardless of the user's display unit; the editor
	/// converts on the way in and out.
	paceSecPerKm?: number;
	runCount?: number;
}

export interface TargetProgress {
	kind: 'distance' | 'time' | 'pace' | 'runCount';
	label: string;
	currentLabel: string;
	targetLabel: string;
	percent: number;
	complete: boolean;
}

export interface GoalProgress {
	goal: RunGoal;
	targets: TargetProgress[];
	overallPercent: number;
	complete: boolean;
	runCount: number;
}

const STORAGE_KEY = 'run_app.goals_v1';

// Serialize to the Android-compatible snake_case wire format so goals
// survive a round-trip through backup/restore and the future settings-bag
// promotion. Aligns with goals.dart's toJson() keys.
function goalToWire(g: RunGoal): Record<string, unknown> {
	const w: Record<string, unknown> = { id: g.id, period: g.period };
	if (g.title != null) w.title = g.title;
	if (g.distanceMetres != null) w.distance_m = g.distanceMetres;
	if (g.timeSeconds != null) w.time_s = g.timeSeconds;
	if (g.paceSecPerKm != null) w.pace_s_per_km = g.paceSecPerKm;
	if (g.runCount != null) w.run_count = g.runCount;
	return w;
}

// Accept both the legacy camelCase keys (written by earlier web versions)
// and the canonical snake_case keys (Android + current web).
function goalFromWire(raw: Record<string, unknown>): RunGoal {
	return {
		id: raw.id as string,
		period: (raw.period as GoalPeriod) ?? 'week',
		title: (raw.title as string | undefined),
		distanceMetres: (raw.distance_m ?? raw.distanceMetres) as number | undefined,
		timeSeconds: (raw.time_s ?? raw.timeSeconds) as number | undefined,
		paceSecPerKm: (raw.pace_s_per_km ?? raw.paceSecPerKm) as number | undefined,
		runCount: (raw.run_count ?? raw.runCount) as number | undefined,
	};
}

export function loadGoals(): RunGoal[] {
	if (typeof localStorage === 'undefined') return [];
	try {
		const raw = localStorage.getItem(STORAGE_KEY);
		if (!raw) return [];
		const list = JSON.parse(raw);
		if (!Array.isArray(list)) return [];
		return list
			.filter((g) => g && typeof g === 'object' && typeof g.id === 'string')
			.map((g) => goalFromWire(g as Record<string, unknown>));
	} catch {
		return [];
	}
}

export function saveGoals(goals: RunGoal[]): void {
	if (typeof localStorage === 'undefined') return;
	try {
		localStorage.setItem(STORAGE_KEY, JSON.stringify(goals.map(goalToWire)));
	} catch {
		/* quota — noop */
	}
}

export function newGoalId(): string {
	return typeof crypto !== 'undefined' && 'randomUUID' in crypto
		? crypto.randomUUID()
		: `g_${Date.now()}_${Math.random().toString(36).slice(2, 8)}`;
}

export function periodStart(
	period: GoalPeriod,
	now: Date,
	weekStartDay: 'monday' | 'sunday' = 'monday',
): Date {
	const d = new Date(now);
	d.setHours(0, 0, 0, 0);
	if (period === 'week') {
		const offset = weekStartDay === 'sunday' ? d.getDay() : (d.getDay() + 6) % 7;
		d.setDate(d.getDate() - offset);
	} else {
		d.setDate(1);
	}
	return d;
}

export function periodEnd(period: GoalPeriod, now: Date, weekStartDay: 'monday' | 'sunday' = 'monday'): Date {
	const start = periodStart(period, now, weekStartDay);
	const end = new Date(start);
	if (period === 'week') {
		end.setDate(end.getDate() + 7);
	} else {
		end.setMonth(end.getMonth() + 1);
	}
	return end;
}

function formatKm(m: number): string {
	return `${(m / 1000).toFixed(1)} km`;
}

function formatMinutes(s: number): string {
	const h = Math.floor(s / 3600);
	const m = Math.floor((s % 3600) / 60);
	if (h > 0) return `${h}h ${m}m`;
	return `${m}m`;
}

/// `secondsPerKm` -> `mm:ss/km` (or "—" if zero / negative).
export function formatPaceSecPerKm(secPerKm: number): string {
	if (!isFinite(secPerKm) || secPerKm <= 0) return '—';
	const m = Math.floor(secPerKm / 60);
	const s = Math.round(secPerKm % 60);
	return `${m}:${s.toString().padStart(2, '0')}/km`;
}

/// Pure evaluator. Given a goal and the full run list, compute progress
/// per active target. Mirrors the shape of `evaluateGoal` in
/// `goals.dart` — active-target list, aggregate percent, overall
/// completion flag.
export function evaluateGoal(
	goal: RunGoal,
	runs: Run[],
	now: Date,
	weekStartDay: 'monday' | 'sunday' = 'monday',
): GoalProgress {
	const start = periodStart(goal.period, now, weekStartDay).getTime();
	const end = periodEnd(goal.period, now, weekStartDay).getTime();
	const inPeriod = runs.filter((r) => {
		const t = new Date(r.started_at).getTime();
		return t >= start && t < end;
	});
	const totalMetres = inPeriod.reduce((s, r) => s + r.distance_m, 0);
	const totalSeconds = inPeriod.reduce((s, r) => s + r.duration_s, 0);

	// Pace calculations exclude cycling — a distance-weighted average
	// would otherwise be dominated by a single long bike ride.
	const paceEligible = inPeriod.filter(
		(r) => (r.metadata?.['activity_type'] as string | undefined) !== 'cycle',
	);

	const targets: TargetProgress[] = [];

	if (goal.distanceMetres != null && goal.distanceMetres > 0) {
		const pct = Math.min(1, totalMetres / goal.distanceMetres);
		targets.push({
			kind: 'distance',
			label: 'Distance',
			currentLabel: formatKm(totalMetres),
			targetLabel: formatKm(goal.distanceMetres),
			percent: pct,
			complete: totalMetres >= goal.distanceMetres,
		});
	}
	if (goal.timeSeconds != null && goal.timeSeconds > 0) {
		const pct = Math.min(1, totalSeconds / goal.timeSeconds);
		targets.push({
			kind: 'time',
			label: 'Time',
			currentLabel: formatMinutes(totalSeconds),
			targetLabel: formatMinutes(goal.timeSeconds),
			percent: pct,
			complete: totalSeconds >= goal.timeSeconds,
		});
	}
	if (goal.paceSecPerKm != null && goal.paceSecPerKm > 0) {
		const paceMetres = paceEligible.reduce((s, r) => s + r.distance_m, 0);
		const paceSeconds = paceEligible.reduce((s, r) => s + r.duration_s, 0);
		const current = paceMetres > 10 ? paceSeconds / (paceMetres / 1000) : 0;
		// Lower-is-better. If we don't yet have any running data, percent
		// is 0; if we beat the target, 100; otherwise (target / current).
		let percent: number;
		let complete: boolean;
		if (current <= 0) {
			percent = 0;
			complete = false;
		} else if (current <= goal.paceSecPerKm) {
			percent = 1;
			complete = true;
		} else {
			percent = Math.max(0, Math.min(1, goal.paceSecPerKm / current));
			complete = false;
		}
		targets.push({
			kind: 'pace',
			label: 'Avg pace',
			currentLabel: current > 0 ? formatPaceSecPerKm(current) : '—',
			targetLabel: formatPaceSecPerKm(goal.paceSecPerKm),
			percent,
			complete,
		});
	}
	if (goal.runCount != null && goal.runCount > 0) {
		const pct = Math.min(1, inPeriod.length / goal.runCount);
		targets.push({
			kind: 'runCount',
			label: 'Runs',
			currentLabel: `${inPeriod.length}`,
			targetLabel: `${goal.runCount}`,
			percent: pct,
			complete: inPeriod.length >= goal.runCount,
		});
	}

	const overall =
		targets.length === 0
			? 0
			: targets.reduce((s, t) => s + t.percent, 0) / targets.length;
	const complete = targets.length > 0 && targets.every((t) => t.complete);

	return {
		goal,
		targets,
		overallPercent: overall,
		complete,
		runCount: inPeriod.length,
	};
}

export function periodLabel(period: GoalPeriod): string {
	return period === 'week' ? 'This week' : 'This month';
}
